### PR TITLE
Fixes #3050 Cannot create rule based on template

### LINKED ISSUE #3050: Cannot create rule based on template
I have this template defined:
```json
[  
   {  
      "uid":"demo:energymeter",
      "label":"Energy Meter",
      "description":"Visualizes the current energy consumption.",
      "configDescriptions":[  
         {  
            "name":"consumption",
            "type":"TEXT",
            "context":"item",
            "label":"Consumption Item",
            "description":"Data source for current consumption",
            "required":true
         },
         {  
            "name":"light",
            "type":"TEXT",
            "context":"item",
            "label":"Color Item",
            "description":"Color light to use for visualisation",
            "required":true
         },
         {  
            "name":"max",
            "type":"INTEGER",
            "label":"Max. consumption",
            "description":"Maximum value (in W) for red light",
            "required":true,
            "defaultValue":1500
         }
      ],
  "triggers": [
    {
      "id": "trigger",
      "label": "Current consumption changes",
      "description": "Triggers whenever the current consumption changes its value",
      "configuration": {
        "itemName": "${consumption}"
      },
      "type": "core.ItemStateChangeTrigger"
    }
  ],
  "conditions": [],
  "actions": [
    {
      "inputs": {},
      "id": "setcolor",
      "label": "Change the light color",
      "description": "Sets the color to a value in the range from green (low consumption) to red (high consumption)",
      "configuration": {
        "type": "application/javascript",
        "script": "var power = trigger.event.getItemState().toString()\nvar percent = power / (${max} / 100) \nif(percent < 0) percent = 0\nvar hue = 120 - percent\nevents.sendCommand('${light}', hue +',100,100')"
      },
      "type": "script.ScriptAction"
    }
  ]
  }
]
```
When creating a rule through the REST API (using the Paper UI), this is what ends up in the rule registry:
```json
[
  {
    "enabled": false,
    "status": {
      "status": "DISABLED",
      "statusDetail": "NONE"
    },
    "triggers": [
      {
        "id": "trigger",
        "configuration": {
          "itemName": "${consumption}"
        },
        "type": "core.ItemStateChangeTrigger"
      }
    ],
    "conditions": [],
    "actions": [
      {
        "inputs": {},
        "id": "setcolor",
        "label": "Change the light color",
        "description": "Sets the color to a value in the range from green (low consumption) to red (high consumption)",
        "configuration": {
          "type": "application/javascript",
          "script": "var power = trigger.event.getItemState().toString()\nvar percent = power / (${max} / 100) \nif(percent < 0) percent = 0\nvar hue = 120 - percent\nevents.sendCommand('${light}', hue +',100,100')"
        },
        "type": "script.ScriptAction"
      }
    ],
    "configuration": {
      "consumption": "Window_FF_Office_Door",
      "light": "Window_FF_Office_Door",
      "max": 1500
    },
    "configDescriptions": [
      {
        "name": "consumption",
        "type": "TEXT",
        "required": true,
        "readOnly": false,
        "multiple": false,
        "context": "item",
        "label": "Consumption Item",
        "description": "Data source for current consumption",
        "options": [],
        "filterCriteria": [],
        "limitToOptions": false,
        "advanced": false
      },
      {
        "name": "light",
        "type": "TEXT",
        "required": true,
        "readOnly": false,
        "multiple": false,
        "context": "item",
        "label": "Color Item",
        "description": "Color light to use for visualisation",
        "options": [],
        "filterCriteria": [],
        "limitToOptions": false,
        "advanced": false
      },
      {
        "name": "max",
        "type": "INTEGER",
        "required": true,
        "readOnly": false,
        "multiple": false,
        "defaultValue": "1500",
        "label": "Max. consumption",
        "description": "Maximum value (in W) for red light",
        "options": [],
        "filterCriteria": [],
        "limitToOptions": false,
        "advanced": false
      }
    ],
    "uid": "rule_1",
    "name": "Energy Meter",
    "tags": [],
    "visibility": "VISIBLE",
    "description": "Visualizes the current energy consumption."
  }
]
```
Note that 
- the placeholders are not replaced with the config values (or is this already a new feature?)
- the rule contains the config params and config descriptions
- the trigger is missing label and description

### PR DESCRIPTION
**Pull request by danchom**
Signed-off-by: Yordan Mihaylov <j.mihailov@prosyst.bg>

### CODE REVIEW

File: bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/RuleEngine.java [hunk 1]
```diff
@@ -275,7 +275,7 @@ protected void removeModuleHandlerFactory(ModuleHandlerFactory moduleHandlerFact
     }
 
     private synchronized void updateModuleHandlerFactoryMap(Collection<String> removedTypes) {
-        for (Iterator<String> it = removedTypes.iterator(); it.hasNext();) {
+        for (Iterator< String>it = removedTypes.iterator(); it.hasNext();) {
```
Comment (kaikreuzer) [inline bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/RuleEngine.java]: is there something wrong with your code formatter...?

Comment (danchom) [inline bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/RuleEngine.java]: I'm using the ESH formatter, but may be its a wrong version.

Comment (kaikreuzer) [inline bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/RuleEngine.java]: could you manually revert those changes?

File: bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/RuleRegistryImpl.java [hunk 2]
```diff
@@ -486,6 +486,10 @@ private Rule resolveRuleByTemplate(Rule rule) {
                 resolvedRule.setName(rule.getName());
                 resolvedRule.setTags(rule.getTags());
                 resolvedRule.setDescription(rule.getDescription());
+
+                // TODO this provide config resolution twice - It must be done only in RuleEngine. Rremove it..
```
Comment (kaikreuzer) [inline bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/RuleRegistryImpl.java]: typo: Rremove

File: bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/RuleEngine.java (hunk 1 above)
Comment (maggu2810) [inline bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/RuleEngine.java]: Shouldn't the Oomph setup grant an up to date formatter?

**Pull request merged**